### PR TITLE
Build and testing for arm64

### LINKED ISSUE #710: Create a CI job for little endian ARM
Ideally we would want to [support Apple Silicon](https://github.com/wasmx/fizzy/issues/709), it seems CircleCI may [not support it anytime soon](https://support.circleci.com/hc/en-us/articles/360047734172-Can-I-build-Apple-Silicon-Universal-apps-). Hence the next best option we have is to run a CI job with Linux and qemu for little endian ARM.

### PR DESCRIPTION
**Pull request by chfast**
Closes #710.

### CODE REVIEW

File: test/testfloat/CMakeLists.txt [hunk 1]
```diff
@@ -26,6 +26,50 @@ if(TESTFLOAT_GEN)
     if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
         # TODO: Clang produces -0 for input 0, see https://bugs.llvm.org/show_bug.cgi?id=47393.
         list(APPEND IGNORE_LIST ui64_to_f64/min)
+    elseif(CMAKE_CROSSCOMPILING AND CMAKE_SYSTEM_PROCESSOR STREQUAL arm64)
+        # These tests fails on arm64 emulation. TODO: Check if NaNs are the reason.
```
Comment (axic) [inline test/testfloat/CMakeLists.txt:30]: @chfast you mentioned these tests are more strict than wasm requires them, but okay on x86. Perhaps mention that here so it is not "misleading" that arm is just failing expected things.

File: test/testfloat/CMakeLists.txt [hunk 2]
```diff
@@ -124,10 +125,10 @@ endif()
 
 function(add_testfloat_selftest NAME ARGS INPUT EXPECTED_ERROR)
     if(INPUT)
-        set(command sh -c "echo ${INPUT} | $<TARGET_FILE:fizzy-testfloat> ${ARGS}")
+        set(command sh -c "echo ${INPUT} | ${EMULATOR} $<TARGET_FILE:fizzy-testfloat> ${ARGS}")
     else()
         separate_arguments(ARGS)
-        set(command $<TARGET_FILE:fizzy-testfloat> ${ARGS})
+        set(command ${CMAKE_CROSSCOMPILING_EMULATOR} $<TARGET_FILE:fizzy-testfloat> ${ARGS})
```
Comment (axic) [inline test/testfloat/CMakeLists.txt:175]: Why not `EMULATOR` here?

File: circle.yml [hunk 3]
```diff
@@ -702,6 +718,9 @@ workflows:
       - spectest:
           requires:
             - coverage-clang
+      - arm64:
+          requires:
+            - coverage-gcc
```
Comment (axic) [inline circle.yml:723]: Why coverage-gcc? That is the only gcc job we have?

File: test/testfloat/CMakeLists.txt (hunk 1 above)
Comment (chfast) [reply test/testfloat/CMakeLists.txt]: I disabled NaN checking for these.

File: test/testfloat/CMakeLists.txt (hunk 2 above)
Comment (chfast) [reply test/testfloat/CMakeLists.txt]: It depends if arguments must be separated with `;` or not.

File: circle.yml (hunk 3 above)
Comment (chfast) [reply circle.yml:723]: Arbitrary ordering. Both use GCC compiler.

File: test/testfloat/CMakeLists.txt [hunk 4]
```diff
@@ -22,63 +22,72 @@ if(TESTFLOAT_GEN)
...
         "f32_trunc\;-rminMag f32_roundToInt"
         "f32_nearest\;-rnear_even f32_roundToInt"
         "f64_ceil\;-rmax f64_roundToInt"
         "f64_floor\;-rmin f64_roundToInt"
         "f64_trunc\;-rminMag f64_roundToInt"
-        "f64_nearest\;-rnear_even f64_roundToInt"
-    )
+        "f64_nearest\;-rnear_even f64_roundToInt")
 
+    string(REPLACE ";" " " EMULATOR "${CMAKE_CROSSCOMPILING_EMULATOR}")
```
Comment (gumb0) [inline test/testfloat/CMakeLists.txt:84]: Can you explain why is this replacing needed?

Comment (chfast) [reply test/testfloat/CMakeLists.txt:85]: Explained.

File: test/testfloat/CMakeLists.txt [hunk 5]
```diff
@@ -22,10 +22,19 @@ if(TESTFLOAT_GEN)
     set(TESTFLOAT_SEED ${TESTFLOAT_SEED} CACHE INTERNAL "testfloat_get seed")  # Save in cache for lookup.
     message(VERBOSE "testfloat_gen seed: ${TESTFLOAT_SEED}")
 
+    # List of test names to ignore.
     set(IGNORE_LIST)
+
+    # Ignore NaN values in _some_ tests. The NaN checks in testfloat are move restrictive
```
Comment (gumb0) [inline test/testfloat/CMakeLists.txt:28]:
```suggestion
    # Ignore NaN values in _some_ tests. The NaN checks in testfloat are more restrictive
```

File: test/testfloat/CMakeLists.txt (hunk 4 above)
Comment (gumb0) [reply test/testfloat/CMakeLists.txt:85]: But why can't it be with spaces from the beginning?

File: cmake/toolchains/arm64.cmake [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+# Copyright 2020 The Fizzy Authors.
+# SPDX-License-Identifier: Apache-2.0
+
+# The toolchain file for arm64 cross-compilation.
+# It selects default version of GCC targeting aarch64-linux-gnu, tested with GCC 10.
+
+set(CMAKE_SYSTEM_PROCESSOR arm64)
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_C_COMPILER aarch64-linux-gnu-gcc)
+set(CMAKE_CXX_COMPILER aarch64-linux-gnu-g++)
```
Comment (axic) [inline cmake/toolchains/arm64.cmake:11]: Shouldn't this file be named `aarch64-linux` perhaps?

**Review by axic — APPROVED** [review]
Apart from those two minor comments I think it looks good.

File: cmake/toolchains/arm64.cmake (hunk 6 above)
Comment (chfast) [reply cmake/toolchains/arm64.cmake:11]: Can be `arm64-linux` if you prefer.

Comment (axic) [reply cmake/toolchains/arm64.cmake:11]: I think it is more correct that way.

**Review by gumb0 — APPROVED** [review]

File: test/testfloat/CMakeLists.txt (hunk 4 above)
Comment (chfast) [reply test/testfloat/CMakeLists.txt:85]: It must be "a semicolon-separated list, then the first value is the command and remaining values are its arguments."
https://cmake.org/cmake/help/latest/prop_tgt/CROSSCOMPILING_EMULATOR.html

File: cmake/toolchains/arm64.cmake (hunk 6 above)
Comment (chfast) [reply cmake/toolchains/arm64.cmake:11]: For the other case, I prefer `arm64` over `aarch64`. It is more informative. Kinda similar issue to `x86-64` vs `amd64`.

**Pull request merged**